lib/encoding.neon

```diff
@@ -0,0 +1,159 @@
+    % This table simply ignores these values.
+    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,    % 0x80 - 0x8f
+    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,    % 0x90 - 0x9f
+    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,    % 0xa0 - 0xaf
+    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,    % 0xb0 - 0xbf
+    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,    % 0xc0 - 0xcf
+    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,    % 0xd0 - 0xdf
+    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,    % 0xe0 - 0xef
+    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,    % 0xf0 - 0xff
+]
+
+%|
+ |  Function: base64Encode
+ |
+ |  Encodes the provided string into Base64 encoded data.
+ |
+ |  Parameters:
+ |      src - The source data to encode.
+ |
+ |%
+FUNCTION base64Encode(src: Bytes): String
+    VAR i: Number := 0
+    VAR encoded: String := ""
+    VAR char1, char2: Number
+    VAR status: Mode := Mode.BYTE1
+    
+    LET source: Array<Number> := src.toArray()
+
+    WHILE i < src.size() DO
+        VAR char: Number := source[i]
+        CASE status
+            WHEN Mode.BYTE1 DO
+                encoded.append(Base64EncodeTable[bitwise.shiftRight32(char, 2)])
+                char1 := bitwise.and32(bitwise.shiftLeft32(char, 4), 0x30)
+                status := Mode.BYTE2
+            WHEN Mode.BYTE2 DO
+                encoded.append(Base64EncodeTable[bitwise.or32(char1, bitwise.shiftRight32(char, 4))])
+                char2 := bitwise.and32(bitwise.shiftLeft32(char, 2), 0x3c)
+                status := Mode.BYTE3
+            WHEN Mode.BYTE3 DO
+                encoded.append(Base64EncodeTable[bitwise.or32(char2, bitwise.shiftRight32(char, 6))])
+                encoded.append(Base64EncodeTable[bitwise.and32(char, 0x3F)])
+                status := Mode.BYTE1
+        END CASE
+        INC i
+    END WHILE
+    
+    CASE (status)
+        WHEN Mode.BYTE2 DO
+            encoded.append(Base64EncodeTable[char1])
+            encoded.append("==")
+        WHEN Mode.BYTE3 DO
+            encoded.append(Base64EncodeTable[char2])
+            encoded.append("=")
+    END CASE
+    RETURN encoded
+END FUNCTION
+
+%|
+ |  Function: base64Decode
+ |
+ |  Decodes the provided Base64 string data into binary data.
+ |
+ |  Parameters:
+ |      src - The source Base64 to decode.
+ |
+ |%
+FUNCTION base64Decode(src: String): Bytes
+    VAR i: Number := 0
+    VAR char1, char2, char3: Number := 0
+    VAR decoded: Array<Number> := []
+    VAR mode: Mode := Mode.BYTE1
+
+    WHILE i < src.length() DO
+        VAR char: Number := Base64DecodeTable[ord(src[i])]
+        IF bitwise.and32(char, 0x80) > 0 THEN
```

Comment:
Using a sentinel value of -1 as above simplifies this test (so you don't have to call bitwise.and32).